### PR TITLE
Clone URL mapping ability

### LINKED ISSUE #4822: clone/install: introduce url mapping convenience?
#### What is the problem?

ATM for a user visiting some datalad-osf published dataset on osf.io it would require either user knowledge or every dataset having a description such as "this component can be git or datalad cloned from a 'osf://ID' URL, where 'ID' is the OSF node ID that shown in the OSF HTTP URL, e.g. https://osf.io/q8xnk/ can be cloned from osf://q8xnk", and then do all url tune up manually to get `datalad` `clone` or `install` working.

### Suggested solution:

I think datalad should make it easier and allow for a simple "copy paste URL I see in the browser bar", thus making `datalad clone https://osf.io/q8xnk/` work as well, while internally mapping it into `osf://q8xnk` for `git clone` operation.

Possible cons:
- if OSF changes their url schema -- would break our mapping.  But given historical stability and presence of those URLs in the wild already, it is very unlikely to happen

Possible implementation:
- datalad internally and also allowing extensions to register new "URL mappers"
- such "mappers" possibly should allow for different/absent context/purpose behavior
  - e.g. for OSF it would only matter for `clone`
  - in a somewhat related (waiting for merge) #4816, where within datalad itself we introduce "mapping" of `shub://` URLs into (minted by singularity hub with timeouts) http urls so that those images could be downloaded from the singularity hub; context would be `download`
- we already hardcode `///` to be a mapping to `http://datasets.datalad.org` -- implementation could generalize it
- some mappers could be as simple as regex based transformation (like in the case of OSF, `///`)

### PR DESCRIPTION
**Pull request by mih**
This implements a simple feature and fix for #4822. It allows for configuring an arbitrary number of URL mapping specification. With this change one can configure datalad to transfer URLs like `https://osf.io/q8xnk/` (copied from a browser window), which do not work, but can be reliably mapped into working URLs (`osf://q8xnk` in this case) via a regular expression specification.

In the example above, this would do the trick:

```
datalad -c 'datalad.url.substitute.osf=,^https://osf.io/([^/]+)[/]*$,osf://\1' clone https://osf.io/q8xnk/
```

as usual, configuration is read from all supported places. Here is an example of a mapping rule
for GitHub

```
% datalad x-configuration
...
# GitHub URL substitution rule
# Substitution specification defined as a string with a match and
# substitution expression, each following Python's regular expression
# syntax. Both expressions are concatenated to a single string with an
# arbitrary delimiter character. The delimiter is defined by prefixing
# the string with the delimiter. Prefix and delimiter are stripped
# from the expressions (Example: ",^http://(.*)$,https://\1"). This
# setting can be defined multiple times. Substitutions will be applied
# incrementally, in order of their definition. The first substitution
# in such a series must match, otherwise no further substitutions in a
# series will be considered. However, following the first match all
# further substitutions in a series are processed, regardless whether
# intermediate expressions match or not.
datalad.clone.url-substitute.github=(',https?://github.com/([^/]+)/(.*)$,\\1###\\2', ',[/\\\\]+,-', ',\\s+|(%2520)+|(%20)+,_', ',([^#]+)###(.*),https://github.com/\\1/\\2')
...
```

It would be possible to implement a mechanism to read additional specifications from an entrypoint that can be provided by extensions. However, it would make also sense to finally support extension-provided (default) configuration and use that mechanism.

TODO:

- [x] Implement sketch for discussion
- [x] Documentation
- [x] Test
- [x] should it rather be `datalad.clone.url-substitute`? (to limit the scope)
- [x] Institutionalize OSF-related mapping from example above?

Closes #4822

### CODE REVIEW

File: datalad/core/distributed/clone.py
Comment (bpoldrack) [inline datalad/core/distributed/clone.py:434]: Re comment - I agree. Don't "fix" explicitly given configs. However, we could use the hint mechanism to point out, that a candidate failed that was the result of such a replacement.

Comment (bpoldrack) [reply datalad/core/distributed/clone.py:434]: To be clear: By agreeing I mean: Do NOT add the original one.

Comment (mih) [reply datalad/core/distributed/clone.py:435]: > Re comment - I agree. Don't "fix" explicitly given configs. However, we could use the hint mechanism to point out, that a candidate failed that was the result of such a replacement.

I am not so sure about that. Config could come from all kinds of places, also sources that are not under a user's control. If any of those substitutions is too loose, what then? File a bug and wait for a release?

Comment (bpoldrack) [reply datalad/core/distributed/clone.py:435]: > If any of those substitutions is too loose, what then? File a bug and wait for a release?

Put a high-priority noop substitution in your confgi? ;-)
Pointing to one issue: If the config is read from everywhere (as it should), explicit priorities seem to be needed (rather than just implicitly by `git-config`-return order.

**Comment by yarikoptic** [comment]
> * should it rather be `datalad.clone.url-substitute`? (to limit the scope)

FWIW - +1 on that. I just wonder if we would have some other 'clone.url' related config options (`.backends`, etc) and might want to make it `datalad.clone.url.substitute` although we do have only few of such 4 level'ed configs and generally they are more of the 3 level form with last level composing into a single word.

**Comment by yarikoptic** [comment]
> > * should it rather be `datalad.clone.url-substitute`? (to limit the scope)
> 
> FWIW - +1 on that. ...

FWIW, I thought that may be we could/should parallel how git does it, e.g. I have
```
[url "git@github.com:"]
    pushinsteadOf = https://yarikoptic@github.com/
    pushinsteadOf = https://github.com/
    pushinsteadOf = http://github.com/
    pushInsteadOf = git://github.com/
```
but trying it out, with 
```
[datalad "url.osf://\1"]
    cloneinsteadOf = "https://osf.io/([^/]*)/?"
```

points to shortcoming of swallowed backslash

```
$> git config -f /tmp/urls.config --get-regex 'url\.*'
datalad.url.osf://1.cloneinsteadof https://osf.io/([^/]*)/?
```

but also if target pattern has `.` in it, I guess splitting apart would also be tricky

```shell
$> cat /tmp/urls.config
[datalad "url.http://something.com/\1/download"]
    cloneinsteadOf = "https://buga.io/([^/]*)/?"

$> git config -f /tmp/urls.config --get-regex 'url\.*'
datalad.url.http://something.com/1/download.cloneinsteadof https://buga.io/([^/]*)/?
```
although we could always assume that it is `datalad.url.<str>.whatinsteadof` and thus `.rsplit(1)` everything after `datalad.url.`

**Comment by mih** [comment]
#### re config style @yarikoptic

I was toying with the idea to adopt that pattern a bit more

```config
[url "osf://\\1"]
        datalad-clone-insteadOf = "https://osf.io/([^/]*)/?"
```

```sh
% git config -l | grep osf
url.osf://\1.datalad-clone-insteadof=https://osf.io/([^/]*)/?
```

(also note the escaped backslash)

I don't see why we could not use the `url.` namespace directly in this fashion, and it would make things look nicer and better grouped, per URL (i.e. github handling, OSF handling ...).

The counter argument is that we need/require(?) a different URL specification style -- which actually makes it a similar, but different beast.

However, which could extend (instead of replacing) the specification, such that if we see something like this:

```
[url "git@github.com:"]
    datalad-clone-insteadof = "https://yarikoptic@github.com/"
```

we replace it internally with:

```
[url "git@github.com:\\1"]
    datalad-clone-insteadof = "https://yarikoptic@github.com/(.*)$"
```

i.e. mimic the base URL replacement done by Git.

#### re priorities @bpoldrack 

> Pointing to one issue: If the config is read from everywhere (as it should), explicit priorities seem to be needed (rather than just implicitly by git-config-return order. I think that is true independently of the question at hand. There could be conflicting substitutions and you don't want to try the wrong one first everytime.

I am struggling to come up with a real use case for this -- if you had some in mind, please share. If there is no real use case, I think we should not implement prioritization. If we find out later that we need it, we could easily extend the setting name scheme ala `datalad.get.subdataset-source-candidate-00...`:

```
[url "osf://\\1"]
        datalad-clone-insteadof-100onlythisworks = "https://osf.io/([^/]*)/?"
```

My rational for "we don't need this" is:

- for subdataset clone candidates we already have a mechanism that can do this -- no need to have two alternatives -- so we are talking about initial clones only
- for initial clones the cost of a failed attempt is much lower (i.e. not attempted wrongly for each of 10k subdataset, but only once for the root)
- we have a prioritization of config sources already, a system/user/repo-provided suboptimal configuration can already by overridden

So Q is: which use case cannot be handled by the levers that are already in place?

**Comment by yarikoptic** [comment]
> I don't see why we could not use the `url.` namespace directly in this fashion, and it would make things look nicer and better grouped, per URL (i.e. github handling, OSF handling ...).

primarily for consistency since all datalad-specific settings reside under `datalad.` .  What I also wonder -- may be should be a hybrid like

```
[datalad "url.datalad-osf"]
    cloneinsteadOf = ",https://osf.io/([^/]*)/?,url.osf://\1"
```
or generally
```
[datalad "url.<mapid>"]
    cloneinsteadOf = "<substitution-pair>"
```
which might help to avoid any possible oddity from sticking regexes into section names, easier overrides (we would need to disallow multiple entries per section I guess), etc?

**Comment by mih** [comment]
> ```
> [datalad "url.<mapid>"]
>    cloneinsteadOf = "<substitution-pair>"
> ```

TBH, I don't really like it. The only thing is has in common with the git-construct is the name. I do like the `<substitution-pair>` syntax better than what git does. Unless there is a good use of the `mapid`, I think I came almost full circle to `datalad.clone.url-substitute = "<substitution-pair>"`.

**Comment by mih** [comment]
I have updated the PR with a new implementation that is capable of providing a fix analog to https://github.com/datalad/datalad/pull/5709, but without custom service dependent conditional. Instead, we can define and ship a default mapping for Github URL, as demo'd in the included unit test at https://github.com/datalad/datalad/pull/5749/files#diff-9789036616f2c57b1a1882ce86a6dc36cd5611c24e73a17eb1305da8335d98d8R1604

A full documentation update is still pending (which is pending a decision of match prioritization, see notes inside), but maybe this is already enough for you @yarikoptic to get an idea where I am going.

In contrast to #5709 this is a fairly generic feature with relatively wide applicability (e.g. #4822), but pretty much any kind of URL sanitization, redirection, reconfiguration. It also makes my suggestion in https://github.com/datalad/datalad/pull/5709#issuecomment-860260616 obsolete, as it offers the same functionality without the custom building block.

**Comment by mih** [comment]
This is stalled by https://github.com/datalad/datalad/issues/5767

File: datalad/core/distributed/clone.py
Comment (yarikoptic) [inline datalad/core/distributed/clone.py:411]: Not sure why longest would be desired

File: datalad/core/distributed/clone.py
Comment (yarikoptic) [inline datalad/core/distributed/clone.py:427]: Debug statement! :heart:

File: datalad/core/distributed/clone.py
Comment (yarikoptic) [inline datalad/core/distributed/clone.py:383]: I guess we would need to introduce `datalad.clone.url-substitute.<series>.priority` or alike?

**Review by yarikoptic — APPROVED** [review]
Didn't try, left comments on TODO comments, but otherwise - love it

**Comment by bpoldrack** [comment]
@mih 

> My rational for "we don't need this" is:
>  -  for subdataset clone candidates we already have a mechanism that can do this -- no need to have two alternatives -- so we are talking about initial clones only
>  -  for initial clones the cost of a failed attempt is much lower (i.e. not attempted wrongly for each of 10k subdataset, but only once for the root)
>  -  we have a prioritization of config sources already, a system/user/repo-provided suboptimal configuration can already by overridden

Generally agree. What I had in mind is admittedly mostly hypothetical. There could be different handlers from different extensions that want to handle URL from some platform. I might then want to curate the order since I know that I'll mostly want A to handle it and only fall back to B if A can't succeed. Something like that. But I guess we can stick with current ways of overriding for now. I'm fine with deferring that until we find it's actually needed.

File: datalad/core/distributed/clone.py
Comment (mih) [reply datalad/core/distributed/clone.py:427]: ;-)

File: datalad/core/distributed/clone.py
Comment (mih) [reply datalad/core/distributed/clone.py:411]: This is what Git does, when deciding which `insteadof` setting to apply. The rational is probably, the longer the match the more specific. In the regex world things will not be that simple, but the analog "more specific before more generic" could still make sense and be implementable.

File: datalad/core/distributed/clone.py
Comment (mih) [reply datalad/core/distributed/clone.py:383]: That, or sorting based on series ID, or inspection of the regex match specification (see "longest match" comment). I think there are multiple possible ways forward.

**Comment by mih** [comment]
Thanks for the comments and the approval. I guess this is good to go now, and we will see when the need for priorities arises.

**Pull request merged**
